### ISSUE TITLE
Bug breaking links: linkUrlsInTrustedHtml does not handle entities correctly

### ISSUE THREAD

**Issue by Llbe** [Art4/urllinker#2]
```
$urlLinker = new Youthweb\UrlLinker\UrlLinker();

var_dump ($urlLinker->linkUrlsInTrustedHtml ("http://foo.bar/baz?a=b&amp;a=b"));
```

The result here is
`string(56) "<a href="http://foo.bar/baz?a=b">foo.bar/baz</a>&amp;a=b"`

The expected result is
`string(56) "<a href="http://foo.bar/baz?a=b&amp;a=b">foo.bar/baz</a>"`

This is a bug since '&' must be encoded in HTML (to avoid conflict with the encoding) and linkUrlsInTrustedHtml() expects valid HTML.

**Comment by Art4** [comment]
Thanks for this issue but I can't identify this as a bug.

`http://foo.bar/baz?a=b&amp;a=b` isn't a valid url, because it is already encoded (i.e with `htmlspecialchars()`).
So this url must be either inside a html tag, not encoded as `http://foo.bar/baz?a=b&a=b` or urlencoded as `http://foo.bar/baz?a=b&amp%3ba=b`.

I've written tests for all these cases, see 8ad91f6ea3539ad38f894ee55c01cb925967ce9a

**Comment by Llbe** [comment]
Thanks for your reply!

It's **not a valid URL** but it is perfectly valid HTML.

The HTML is `http://foo.bar/baz?a=b&amp;a=b`

The URL is `http://foo.bar/baz?a=b&a=b`

And the result should be `<a href="http://foo.bar/baz?a=b&amp;a=b">foo.bar/baz</a>`. You are avoding the double encoding already, in escapeHtml().

This is the only way if accepting valid HTML in linkUrlsInTrustedHtml().

**Comment by Art4** [comment]
This would be valid HTML if it was already inside a HTML tag like `<a href="http://foo.bar/baz?a=b&amp;a=b">...`. But it isn't, so it is either valid HTML nor a valid url.

In this case you have to use `$urlLinker->linkUrlsAndEscapeHtml('http://foo.bar/baz?a=b&amp;a=b')` what works fine. See tests in 24924746170ff63b8e0aa52290979678170b70e0.

**Comment by Llbe** [comment]
So you are saying that you can only use `&amp;` inside an attribute in HTML?

Eg the following would **not** be valid HTML: `<p>http://foo.bar/baz?a=b&amp;a=b</p>`?

**Comment by Art4** [comment]
> Eg the following would not be valid HTML: `<p>http://foo.bar/baz?a=b&amp;a=b</p>`?

Sure, this is valid HTML, but without a valid url.

The question is less if this HTML is valid but if it is trusted. So if this `<p>http://foo.bar/baz?a=b&amp;a=b</p>` came from as a request from a user it is not trusted and you should use `$urlLinker->linkUrlsAndEscapeHtml()`.

But if this came from a trusted source than you can use `$urlLinker->linkUrlsInTrustedHtml()` but you can see that it contains not a valid url.

This is a valid url: `http://foo.bar/baz?a=b&a=b`
This is not a valid url: `http://foo.bar/baz?a=b&amp;a=b`

**Comment by Llbe** [comment]
The encoding **are layers**. The `&amp;` belongs to HTML as you know. It has nothing to do with the URL. This is why it still is a valid URL **given that the HTML is interpreted**.

I am not saying that `http://foo.bar/baz?a=b&amp;a=b` is a valid URL. I am saying that it is valid HTML that contains a valid URL.

**Comment by Llbe** [comment]
The linkUrlsInTrustedHtml() should understand HTML correctly and _trusting_.

**Comment by Art4** [comment]
> I am not saying that `http://foo.bar/baz?a=b&amp;a=b` is a valid URL. I am saying that it is valid HTML that contains a valid URL.

If you insert an invalid URL the library will try to recognize as much as possible and put it in a `<a>`-tag. You can't expect that the library will know the meaning of an invalid url.

> The linkUrlsInTrustedHtml() should understand HTML correctly and trusting.

So how should the library decide how to handle an invalid url? The meanings of `http://foo.bar/baz?a=b&amp;a=b` could be:

| valid url  | in `<a href="` |
|--------|---------------|
| `http://foo.bar/baz?a=b&a=b`              | `http://foo.bar/baz?a=b&amp;a=b` |
| `http://foo.bar/baz?a=b&amp%3ba=b` | `http://foo.bar/baz?a=b&amp;amp%3ba=b` |
| `http://foo.bar/baz?a=b%26amp%3ba=b` | `http://foo.bar/baz?a=b%26amp%3ba=b` |

The library will recognize an url until the first query part `http://foo.bar/baz?a=b` and ignore the rest. We can discuss if the library should recognize the url until `http://foo.bar/baz?a=b&amp` because this is also valid. But the rest `;a=b` is invalid and is ignored.

I'd be glad if you would create a PR with a solution how to decide what the library should do.

**Comment by Art4** [comment]
Another example: If you call the url `http://foo.bar/baz?a=b&amp;c=d` PHP will interpret the querystring as:
```
array(2) {
  ["a"]=>
  string(1) "b"
  ["amp;c"]=>
  string(1) "d"
}
```

This is because the url will be urlencoded in the browser and became `http://foo.bar/baz?a=b&amp;amp%3ba=b`.

But how should the library if it should use `urlencode()` or turn `&amp;` back to `&`? You provided "trusted HTML" to `$urlLinker->linkUrlsInTrustedHtml()` so it will not change it but search for valid urls.
But if you put this invalid url into `$urlLinker->linkUrlsAndEscapeHtml()` it will escape the HTML and will recognize the `&amp;` as `&`.

So you have two options:

1. Make sure you insert valid urls
2. Use `$urlLinker->linkUrlsAndEscapeHtml()`

**Comment by Llbe** [comment]
> I am not saying that http://foo.bar/baz?a=b&amp;a=b is a valid URL. I am saying that it is valid HTML that contains a valid URL.
If you insert an invalid URL the library will try to recognize as much as possible and put it in a <a>-tag. You can't expect that the library will know the meaning of an invalid url.

I'm sorry but you are missing the point. I am not inserting an invalid URL. **I am inserting valid HTML**. When the HTML is interpreted, it becomes a valid URL.

The fundamentally important thing is to encode/decode _per context_. You must not handle HTML entities as a part of a URL. Currently you are converting HTML → text without a HTML decode.

Basically linkUrlsInTrustedHtml() should HTML decode before calling linkUrlsAndEscapeHtml() - because the latter method expects text and not HTML.

For this to work the regex in linkUrlsInTrustedHtml() must include HTML entities when finding the URLs.

**Comment by Llbe** [comment]
You are mixing different contexts.

**Comment by Llbe** [comment]
Use html_entity_decode().

**Comment by Art4** [comment]
I'm sorry but I don't get your problem. Can you provide me a specific example of your problem?

If you have the input from above:

```php
$urlLinker = new Youthweb\UrlLinker\UrlLinker();

var_dump($urlLinker->linkUrlsInTrustedHtml("http://foo.bar/baz?a=b&amp;a=b"));
```
 
you will get

| input | output from `linkUrlsInTrustedHtml()` |
|------|----------------------------------------|
| `http://foo.bar/baz?a=b&amp;a=b` | `<a href="http://foo.bar/baz?a=b">foo.bar/baz</a>&amp;a=b` |
| `<a href="http://foo.bar/baz?a=b&amp;a=b">foo.bar</a>` | `<a href="http://foo.bar/baz?a=b&amp;a=b">foo.bar</a>` |
| `<p>http://foo.bar/baz?a=b&amp;a=b</p>` | `<p><a href="http://foo.bar/baz?a=b">foo.bar/baz</a>&amp;a=b</p>` |

> Use html_entity_decode().

Why should it use `html_entity_decode`? You inserted *trustful* HTML. But you can use `html_entity_decode()` on your own before insert it to `linkUrlsInTrustedHtml()`.

**Comment by Llbe** [comment]
I don't want to be rude but you should study how to work with encodings. I promise you it will prove beneficial.

The problem with linkUrlsInTrustedHtml() specifically is that links are not extracted _including_ entities.

You cannot decode before inserting - because you cannot decode before "parsing"/regex.

Because regex is used (and not libxml) - you need to extract links including entities **and**

 a) decode the entities before inserting to linkUrlsAndEscapeHtml()

 or

 b) as today: avoid double encoding in linkUrlsAndEscapeHtml()


..

As a side note, ideally you should only extract links from HTML including entities that are within a link and not at the end of a link (or at the beginning of a link).

This is in order to handle a link such as (text): `<http://foo.bar/baz>`, which in HTML would be `&lt;http://foo.bar/baz&gt;`, where the link should be `http://foo.bar/baz`.

However `http://foo.bar/baz&gt;baz` in HTML would be interpreted as link `http://foo.bar/baz>baz`.

('<' and '>' is allowed unencoded in URLs, see RFC 3986 - it is considered unsafe/unwise by RFC 1738).

I take this as an example because '<' and '>' is more obviously problematic in HTML. Technically, '&' is allowed unencoded (it is only so called ambiguous ampersands that are not allowed - ie ampersands that can be interpreted as an entity but that are not a valid entity).

The '<' '>' case are not as important as the `&amp;` case - since that WILL break common links in valid HTML.

I have a working and secure handling of user input in my application. I use HTML for rich text formating. I want to convert links on the fly because how links are converted depends (I have a vary on target and rel attributes). This library does not work because HTML is not handled correctly. Otherwise, it seems to be a good improvement on the kwi original.

**Comment by Art4** [comment]
Thank you for your patience. :blush: I'm always interested in learning new things. I am far away from being an expert on encoding, but I try to understand things before I can improve them. So I am really glad for your issue and I will work through your comment and discuss this issue internally.

> This library does not work because HTML is not handled correctly. Otherwise, it seems to be a good improvement on the kwi original.

Thank you. And we are interested in improving this library and are thankful for every contribution.

**Comment by Art4** [comment]
Thanks for this issue and for the patience you had explaining the nuances of encoding here. :blush: Your comments taught me a lot about how `&amp;` in trusted HTML belongs to the HTML, not to the URL, while the *meaning* of the URL is the decoded value. I understand this problem much better now: when a character reference appears inside a link, the link should point where the HTML says it points, which is the decoded URL.

With #40, the examples from this discussion now produce:

- `http://foo.bar/baz?a=b&amp;a=b` gets linked as `<a href="http://foo.bar/baz?a=b&amp;a=b">foo.bar/baz</a>`
- `<p>http://foo.bar/baz?a=b&amp;a=b</p>` becomes `<p><a href="http://foo.bar/baz?a=b&amp;a=b">foo.bar/baz</a></p>`
- `<a href="http://foo.bar/baz?a=b&amp;a=b">foo.bar</a>` is left untouched, because it is already a link

The exact case from your side note works as well: a reference to a URL character (`&#38;`, `&#x26;`, and so on) is part of the URL, while a reference to a character that can't appear in a URL (`&lt;`, `&gt;`) still flanks it. And your `&lt;http://foo.bar/baz&gt;` example correctly links only `http://foo.bar/baz`.

The fix is in #40 and will go out with 2.2.0. Two follow-ups:

1. The old behavior, splitting every URL at a character reference, is still available behind the new option `cutUrlsAtEntities` as a migration aid. We plan to deprecate it in 2.3.0 (#38) and remove it in 3.0.0 (#39).
2. While implementing this, we also found that a trailing reference at the very end of a URL was being split in half, because its terminating `;` was swallowed as punctuation. #40 fixes that too.

I'd be glad to hear if the new behavior works for your use case.
